### ISSUE TITLE
Increase Video Processing from 1 Frame to Multi Frames Intelligently

### ISSUE THREAD

**Issue by hollowaykeanho** [hollowaykeanho/Upscaler#1]
# Description
```
Please provide a short description of what feature you're looking forward to
enhance below. Please include the story behind your idea as well to give a
better visualization of your idea.
```
Basically to speed things up.

# Expected Behavior
```
Please specify the expected behavior of your requested enhancement. Some great
and helpful pointers are your expected interface (e.g. command patterns,
simple sketches of the user interface, etc).
```
Upscaler can work a lot faster via work.

# Current Behavior
```
Please specify the current behavior (e.g. workaround, blockage, etc).
```
Upscaler only work on 1 frame at a time.

# Attachment
```
Please drag and drop the necessary data files (e.g. screenshot, logs, etc)
below.
```

**Comment by hollowaykeanho** [comment]
Help needed as I do not have a better graphics card. Limited to only `NVIDIA GeForce MX150` from a laptop to test its working capability on consumer products.

**Comment by hollowaykeanho** [comment]
Not possible with POSIX and Powershell. Neither of them provides some kind of sync mechanism like mutex lock or channels. Closing this issue as won't fix.

**Comment by hollowaykeanho** [comment]
Re-open as now we have a possibility to process frame upscaling concurrently using scoring system.

**Comment by hollowaykeanho** [comment]
@Joly0, @JeanShuralyov , can I acquire `--parallel` argument name for experimenting this feature ya?

like `--parallel 2` for 2 images simultenously.

**Comment by Joly0** [comment]
Sounds good for me, i´d have to adjust the powershell side of this when you are done

**Comment by hollowaykeanho** [comment]
TQ & noted. I'll begin the enhancement work after clearing some works at hand.

**Comment by JeanShuralyov** [comment]
@hollowaykeanho , fine by me. Will you be using my fork repo for this enhancement?

**Comment by hollowaykeanho** [comment]
yes. Please keep it.

**Comment by hollowaykeanho** [comment]
@Joly0 , I might need help with your graphics card. =x

Appearently, I completed my prototyping codes but same like my previous attempt, I blew my vRAM when parallel is set to 2. (It demands 2400MB I believe).

![2023-07-13-19-50-28](https://github.com/hollowaykeanho/Upscaler/assets/7259572/2f2aefce-5fd6-4de7-8c81-35eaceb038de)

I will complete the commit and push to Jean's forked repo for easier referencing.

**Comment by Joly0** [comment]
Sure, how can i help?

**Comment by hollowaykeanho** [comment]
Okay. I just pushed in parallel processing capability for `init/unix.sh` side of things. Diff: https://github.com/hollowaykeanho/Upscaler/commit/aa4fabb18a8256159ccb5555f787f06546c33536

You can try it out with `--parallel 2` (I can't get through due to the hardware limitation). Once translated to the PowerShell side, this issue can be marked completed. =)

**Comment by JeanShuralyov** [comment]
@hollowaykeanho , there is no way my Iris can operate `--parallel 2` if your Nvidia failed. Passed. Otherwise, great job.

**Comment by hollowaykeanho** [comment]
noted. In that case I won't increase its parallelism for benchmark testing then since we do not have proper hardware to test it iteratively.

**Comment by hollowaykeanho** [comment]
@Joly0 , is it possible to translate the current outcome ya? It will be the last one since I'll drilling into Tencent NCNN framework for the next step.

**Comment by Joly0** [comment]
Hey, yes, i am on it getting this done in powershell, but i dont have that much time currently, so i cant tell, when i am finished. I will try my best to get this done asap.

**Comment by hollowaykeanho** [comment]
Haha.. no problem. It's open-source so no deadline. Take your time.

In that case I will begin the work on the source codes now. Let me know when you're ready for release.

**Comment by hollowaykeanho** [comment]
Just in case: please note that you're not required to create directory for each frames that I used.

I only did that because `mkdir` can behave safely in Linux side. You might need to find out is PowerShell has a similar command that has a safe concurrent mechanism.

**Comment by Joly0** [comment]
Btw, in powershell this is a hell of a ride, but i am getting quite close i guess, but i have to set powershell 7 as a requirement for this to work. Powershell 5 has a few techniques to make parallel execution working (jobs, runspaces, workflows), but none of them are easy to work with.
I could try to get it working with jobs or workflows, but its such a pain to work with functions and variables outside of a job or workflow scope. 
Powershell 7 has the "foreach-object -parallel " function, which works atleast better. Still having issues with variables and functions, but its atleast not too difficult to work with.

So if its ok, i will get this working with powershell 7, otherwise if you want, i´ll adjust it to powershell 5

But just so you see it, its working
![image](https://github.com/hollowaykeanho/Upscaler/assets/13993216/ea5cbf2b-df66-444d-a8d3-d2e37a3c044b)
(running the benchmark here with --parallel 2)

**Comment by hollowaykeanho** [comment]
> Powershell 5 has a few techniques to make parallel execution working (jobs, runspaces, workflows), but none of them are easy to work with.

Hahaha.. completely agree with you. I'm currently working on a PowerShell + POSIX compliant Shell CI infra at https://github.com/ChewKeanHo/AutomataCI/tree/experimental. Currently working on the Docker image packaging job development. 

Once I completed the AutomataCI to a stable release, I will port it here and there will be various released packages (for sure having `.deb`, `.rpm`, `.flatpak`, `.tar.xz`, and `.zip`. 


You can make use of my newly developed shell-based libraries here: https://github.com/ChewKeanHo/AutomataCI/tree/experimental/automataCI/services

> (running the benchmark here with --parallel 2)

Good luck! I couldn't run here. 

> I´ll let this run through and will try with some other files aswell and i have to get the control file working again, because it is borked now XD

Haha. Yeap. Some changes done because the execution is different now.

**Comment by Joly0** [comment]
Btw, just to give a quick update. I havent had any time in the past few months to look into this at all, which i am quite sorry for. But i think i´ll might have some time now to get this a all fixed, finish it up and push it out, so it can be merged :D

**Comment by corygalyna** [comment]
@Joly0 , I had cleaned up the project dashboard into a road-map kanban style already. This should further the repository development with a more directional approach.

*Yawn*. Enjoy. Oh by the way, @hollowaykeanho , did something mad in this project: https://github.com/ChewKeanHo/AutomataCI/tree/main/automataCI/services

I believe controlled multi-threading is available in there `io/sync.*`

**Comment by Joly0** [comment]
https://github.com/Joly0/Upscaler/commit/9ce20e919df387902eb9bd3c77b29c8eb776c582

**Comment by hollowaykeanho** [comment]
I assume it was tested right? In that case,  I will merge it in.

**Comment by Joly0** [comment]
Apparentely the retry function is not yet working correctly, but otherwise yes, this works. I can do some testing again later and can check how much of a difference this makes going from parallel 1 to 2 or 4.

Also while we are at it, it might be useful to have a shorter second benchmark/test video for people to check the performance of parallel easier. It can take quite a while for even the 400 frames benchmark video to finish. I think just a second of video at 25fps should be enough

**Comment by hollowaykeanho** [comment]
> Also while we are at it, it might be useful to have a shorter second benchmark/test video for people to check the performance of parallel easier. It can take quite a while for even the 400 frames benchmark video to finish. I think just a second of video at 25fps should be enough

All right. I will look into it.

**Comment by hollowaykeanho** [comment]
I merged your patch into the production already: 3ad9974a70eb8aa998579689741ff3b6b7ecee96

The test video will need to generate a new one. I will generate a test result so that it's clearer.

**Comment by NoDiskNoFun** [comment]
Hey there, are there any tester needed? I have a Manjaro System with to RTX 3060 and one RTX 3080 inside. I would love to help out here :)

**Comment by hollowaykeanho** [comment]
Haha, thank you. A benchmark data on the hardware would be helpful (guiding people) too. Here's the section on `README.md`:

https://github.com/hollowaykeanho/Upscaler?tab=readme-ov-file#benchmarks

You'll need the `tests` directory. Inside, there is a `benchmark.cmd` that will generate the stats against `tests/video` and `test/image` samples.

**Comment by hollowaykeanho** [comment]
Hi @NoDiskNoFun! Sorry for the late reply.

> I started by considering how to add that feature, but eventually ended up rewriting unix.sh almost entirely 😄 You can check out my version here:

There is actually an on-going re-write inside `init/` directory (start script is `init/start.sh.ps1` pointing to `start.sh` and `start.ps1`). You're actually running the old prototype code. It's just we both don't have much time to continue working on it (yet). You can try running it `$ ./init/start.sh.ps1 --help`. So far I only implement only image upscale. Currently working on HestiaKERNEL + HestiaFFMPEG clean up before continue to video upscale.

FYI, I'm already cleaning up the library inside `init/services/` directory (those with prefix `Hestia`). However, I need to strike multiple goals in 1 stone so Hestia libraries has to be precise and clean so that we don't need to re-visit all these primitive functions in the future.

Hmm, let me clean up the existing codes (the start script contains my unit-test hacking) for you since you're working on it.

> My version doesn’t process everything in two distinct phases (extract → scale). Instead, it splits the work into jobs, each handling a fixed number of frames (default: 50). Each job performs both extraction and 
, and jobs can run in parallel — which massively improves performance.

We managed to develop a shell based sentinel multi-threading already. The library codes are in `init/services/HestiaKERNEL/Parallelism` (note: will change to plural in next commit since Hestia project already got its standard). On UNIX shell, we make use of `mkdir` because POSIX shell does not have mutex sync and `mkdir` has the ability to fill in the role.

However, our last test on laptops and hardware constrained devices cannot upscale in parallel beyond 2 frame per transient and we have to support them for general public use (not everyone has super high-end hardware) - as in `Function always first before performance`.

The design was later pivoted to using `--parallel` flag to allow user to designate number of parallel job based on the hardware he/she owned.

The hard part wasn't about implementing the missing parallelisms for shell but to make sure both PowerShell and Shell APIs are completely aligned (so that Joly0 and I can directly communicate in sync despite we're from different languages).

> Support for RIFE to allow AI frame interpolation (doubling the frame count) (BETA)

Focus on this. This is good. =) Looking forward to it.

> I’d love to get your feedback on this approach, and possibly discuss integrating parts of it back into the original project.

Feel free to mess with the project in your fork repository. You can open your issue ticket over there at mention me. I'll drop-by and answer back.

For now, other than answering your query, I can't commit much yet. Been urgently rushing a data OS to secure most of data scattered across the cloud for mass data consolidation (if you notice my GitHub activity chart had significantly dropped). Once I'm done, I'll come back.

The pull request policy is that:

1. I will review the code and merge in if the code is readable.
2. The only requirement is the thing you're implementing is functional.
3. Normally I will git pull your foked repo and cherry-pick your committed patches. I will accept pull request for GitHub badge sake (remember to place `Co-Authored by:` with 1 of your friend or me or joly0 - see commit history). However, it will get overridden mainly because of git signature (I will sign on top of it for you).
4. Then I will amend code hygiene in a new commit if I spotted some.

> Huge thanks for your awesome script — without that solid base, I wouldn’t have had nearly as much fun diving into this rewrite!

I thanks for your patience and efforts. Looking forward to merge your changes and welcome aboard! 


> My version doesn’t process everything in two distinct phases (extract → scale).


The extract step phases is mainly to avoid race conditions on extract and merge for parallelism. It is also to detect which phase is the one that provide the error. The design is that both phases should operate with the `--parallel` flag independently on their own.

For what I know, I don't recall ffmpeg can do multi-write so those 2 outer phases have to be done serially. FFMPEG wasn't the timing consuming part; it's the upscaling frame-by-frame.


> Just as a quick note: since my English isn’t great, I used AI assistance to write this message (based on my own ideas and input) — of course I reviewed everything myself before sending.

English, Chinese (简体, 繁体), Malay, French, and a little German (I'm way to elementary for this one). =) Don't worry to much. You're doing great. This is actually a fun project and to extend one's ability to upscale media.

Site-note: thanks for helping out in other ticket. Saw your email yesterday. =)

**Comment by NoDiskNoFun** [comment]
Guten Tag und Bonjour! Thanks so much for the detailed reply — and no worries about the delay!

I took a quick look at the code under `init/start.sh`, and wow — that's a whole different beast compared to unix.sh. I think I’ll need more than just a text editor for that one 😅 Also when it starts up, it kinda feels like I’m entering the Matrix! 🔴 lol
Out of curiosity, when you mention Hestia, are you referring to this project? → https://github.com/ChewKeanHo/HestiaKERNEL

> However, our last test on laptops and hardware constrained devices cannot upscale in parallel beyond 2 frame per transient and we have to support them for general public use (not everyone has super high-end hardware) - as in Function always first before performance.

Regarding parallelism: yeah, you’re absolutely right. Functionality before performance makes total sense, especially for users on less powerful machines. (Though in my defense — my “super hardware” kinda just happened 😄)

> The design was later pivoted to using --parallel flag to allow user to designate number of parallel job based on the hardware he/she owned.

My version of the script also allows configuring the number of extractors and upscalers separately — each runs as an independent process. I do realize I still need to update the README to document all the new parameters properly. Will add that soon!

> The extract step phases is mainly to avoid race conditions on extract and merge for parallelism. It is also to detect which phase is the one that provide the error. The design is that both phases should operate with the --parallel flag independently on their own.

I definitely fought my share of race conditions too — and (mostly) lost, lol. That said, most of the time the subprocesses (aka jobs) behave nicely: they usually don’t get in each other’s way and I manage to keep the number of parallel extract and upscale workers within the user-defined limits. Only occasionally, one extra worker sneaks in somehow, or the script struggles to reliably detect which GPU is actually free for the next job — race conditions can be sneaky like that. In my script, the jobs all run within a trap, so a failure in one should terminate the main process as well. It works most of the time — but not always — so that’s definitely something I want to get to the bottom of!

> FFMPEG wasn't the timing consuming part; it's the upscaling frame-by-frame.

In my testing, FFmpeg starts slowing down massively once you hit tens of thousands of frames. In one of my test files, I noticed that FFmpeg gradually slowed down — from about ~3 FPS at the beginning to around ~0.3 FPS by the time it reached the 10,000th frame. From what I could find, this seems to be due to FFmpeg having to decode all prior frames up to the requested one. That said, things start to look very different when you throw three GPUs at the problem — figuratively, of course; I wouldn’t actually throw my graphics cards. 😅

That’s why I was particularly curious about the comment in your code around line 571:

`# generate each frame selectively for highest quality extraction.`
Do you happen to have more details on that? I couldn’t find anything in FFmpeg docs that suggested different extraction methods would affect frame quality. If there’s a nuance here I missed, I’d love to understand it better!

By the way — I'd actually been thinking about turning my fork into a standalone project, more focused on high-end setups and large movie-sized videos. I’d likely drop Windows support in the process to simplify things. That said, since my version still builds on your work, I wanted to ask if you’d be okay with me doing that — I want to make sure I show proper respect for the effort and foundation you've created. Of course, I’d still love to help out here and learn more about open source development in general. The PR I submitted yesterday was literally my first one ever, so I’ve got a lot to learn. 😄
In that process, I’ll also try to rework my RIFE implementation in a way that makes it easier to eventually merge back. I’ll need to do some research to figure out the best way to structure that — but it’s definitely on my list.

Plus, I'd absolutely be happy to contribute to the German language support — it’s my native language, and I noticed there’s already a ticket open for that. Let me know how I can help.

Thanks again for taking the time to respond so thoroughly. I really appreciate your kind and thoughtful reply — it’s a welcome change, as I’ve unfortunately received a lot of negative responses in the past. 
This project has been a ton of fun to dive into, and I’m looking forward to continuing to learn from you and contribute where I can.

**Comment by Joly0** [comment]
Hey, i wanted to provide my feedback here aswell. So in general the approach using separate jobs for both parts (extract -> scale) makes sense, although i am not 100% setting a fixed value of 50 is the best approach. This should really be an issue, but it could happen that one job is much much faster than a second one, and the whole process would have to wait for the second job to finish, a more dynamic approach might be better. On the other hand, i do not think this is a big issue, as 50 frames in a video of 1.000.000 frames is next to nothing, still might be a good idea to look at this (or use a smaller default value, not sure about this). Maybe i am overthinking this and this is basically a non-issue 😀

Otherwise, great job from what i can see, this looks really good. I would have a good work to do to get this properly implemented in powershell aswell, to make both of them allign. I am not sure how i would best handle this, could be a tough task, but i think thats nothing anyone of you guys need to worry about 😄

Btw, regarding language or multi language to be more specific, we should make a plan how we will implement and use this so others can easily add new languages without needing to add any code or so. So general resource files might be a possible idea, but this would need to be implemented for all configurations (linux, mac and windows) in the same way so new languages and changes are properly handled everywhere. Not sure if @hollowaykeanho already started with this somewhere, i havent seen anything yet regarding multi language support.
Also i am german aswell, so we should definitely talk about who at the end will implement the german language, so we do not both start with it, which would be duplicate work 😅 Rock paper scissors or so? The winner can do german language 😂

**Comment by NoDiskNoFun** [comment]
Hey, thanks for your reply, too. Its really nice to have such a warm welcome here :) The jobsize of 50 frames is just default if the parameter --max-job-size is not given. I played around with this value a bit and its really specific to the total frames of the given video file. For example for the sample video with 400 Frames i started my script with three upscaling workers with a job size of 134 to evenly distribute the load over my three GPUs. (BTW that took just 172 seconds to finish). But i guess if one think long enough about this there surely is a system to calculate the best jobsize per file.

> Otherwise, great job from what i can see, this looks really good.
Thanks a lot for your feedback! But i have to admit that there are still some bugs and i took some shortcuts to make things working i didnt fully understand (at least for now). For example the extracted frames will always be png, because i hardcoded it. 

I already thought about multilangues support a bit. I think a good approach (at least for bash) would be removing all hardcoded text messages from the script and replace them with variables. Then we could create languagesfiles which could be sourced. Something like this:

lang-file-de

thisisamessage="Dies ist eine Nachricht"

inside unix.sh

source lang-file-de

_print_status error $thisisamessage

but this is getting a bit off topic. I didnt see an open ticket for general language support. May i open one to discuss this matter there?

@Joly0: I chose rock  xD

Just a quick note: This was me, not the AI. If you did understand me, great! :D

**Comment by hollowaykeanho** [comment]
Hi all. Sorry for the late reply.

> Out of curiosity, when you mention Hestia, are you referring to this project? → https://github.com/ChewKeanHo/HestiaKERNEL

Yeap. That's the GitHub housing repository. I'm working inside my local Gitea here (haven't push it out yet due to code cleanliness). You can look into this repository's `init/services` directory here (the local copy). I'm clearing the `HestiaKernel`. This library will be renamed soon cuz I finally found the best suited for this architecture (go beyond Shell and Powershell).

Yeah, it's a complete re-write. Hestia libraries solve a lot of repeating problems here. 1 shot, many checklist done.

> Regarding parallelism: yeah, you’re absolutely right. Functionality before performance makes total sense, especially for users on less powerful machines. (Though in my defense — my “super hardware” kinda just happened 😄)

Don't worry. We do want powerful candidates. ;-) 

Site-note: Can you setup your sponor link and provide me your name + email please? Wanna credit you.

> I definitely fought my share of race conditions too — and (mostly) lost, lol. That said, most of the time the subprocesses (aka jobs) behave nicely: they usually don’t get in each other’s way and I manage to keep the number of parallel extract and upscale workers within the user-defined limits. Only occasionally, one extra worker sneaks in somehow, or the script struggles to reliably detect which GPU is actually free for the next job — race conditions can be sneaky like that. In my script, the jobs all run within a trap, so a failure in one should terminate the main process as well. It works most of the time — but not always — so that’s definitely something I want to get to the bottom of!

The secret to great parallelism is you can see the process path and data path clearly at any transient moment. There are 2 common strategies ('Sentinel' vs. 'Divide & Conquer'). In your 50 jobs, you're deploying 'Divide & Conquer' (quite straight forward). 'Sentinel' is having a thread to watch all over others (kinda like 1 manager with many workers). 

Basically you can do thought experiment like: 10000 people (your CPU threads) is requesting for entry from just 1 counter (video file) at a theme park. There's only such much that 1 counter can do.

Once the images are separated. Divide and conquer. Go wild with full might. If we continue the thought experiment, this is where that 10000 people each handles 1 separate game they can find/deal. =)

Same case as 'merge back' with 'extract' - only and strictly 1 write thread. However, we don't need to worry because ffmpeg handles it internally. If we continue the thought experiment, now everyone wants to go home, but via same 1 counter.

If at any given transient point at any path you can't 100% confidently predict what would happen, that's a hint parallelism condition is not ready. Try investigate with this thought experiment see it helps.

> The PR I submitted yesterday was literally my first one ever, so I’ve got a lot to learn. 😄

Haha, take your time and experiment. Don't worry. I got a local backup copy. =p


> In my testing, FFmpeg starts slowing down massively once you hit tens of thousands of frames. In one of my test files, I noticed that FFmpeg gradually slowed down — from about ~3 FPS at the beginning to around ~0.3 FPS by the time it reached the 10,000th frame. From what I could find, this seems to be due to FFmpeg having to decode all prior frames up to the requested one. That said, things start to look very different when you throw three GPUs at the problem — figuratively, of course; I wouldn’t actually throw my graphics cards.

Wait what?! This sounds like FFMPEG bug. Per its manual it suppose to seek that 1 particular frame immediately and extract out; not scan each of them from the start. It shouldn't always start from 0.

Need to double confirm outside of this project.


> Do you happen to have more details on that? I couldn’t find anything in FFmpeg docs that suggested different extraction methods would affect frame quality. If there’s a nuance here I missed, I’d love to understand it better!

This means ffmpeg will probe the video for its dimension and resolution then extract the maximum possible quality from it as original as possible. ffmpeg can modify image frame at will which is not something people want (intuitively). Last I did was a bunch of research and read up a computer vision professor's blog (sorry, lost it to garbage SEO in this AI era zzzz).

I do a fresh research again and these I find still ok (haven't try). I normally just read `--help` and experiment. XD


References:

1. https://nulldog.com/extract-high-quality-jpeg-from-video-with-ffmpeg 
2. https://superuser.com/questions/1552112/ffmpeg-extract-every-frame-from-a-video-no-quality-loss-and-resize-shorter-s


> Not sure if @hollowaykeanho already started with this somewhere, i havent seen anything yet regarding multi language support.

Yes, already got 3 solutions. Problem is the library architecture.

> Plus, I'd absolutely be happy to contribute to the German language support — it’s my native language, and I noticed there’s already a ticket open for that. Let me know how I can help.

> Also i am german aswell, so we should definitely talk about who at the end will implement the german language, so we do not both start with it, which would be duplicate work 😅 Rock paper scissors or so? The winner can do german language 😂

German is 1000% in the list.

----

Alright Alright. I got the message.

Let me up my game now and layout the vision. Then, everyone can help without bottlenecking at me. I'll get it done by this week.
